packages/node-core/src/indexer/poi/poi.service.ts

```diff
@@ -1,55 +1,261 @@
+          queries.push(`UPDATE ${tableName} SET hash = NULL;`);
+          queries.push(
+            `CREATE UNIQUE INDEX IF NOT EXISTS "poi_hash" ON ${tableName} ("hash") WHERE "hash" IS NOT NULL`
+          );
+        }
+        if (!checkResult.parent_nullable) {
+          queries.push(`ALTER TABLE ${tableName} ALTER COLUMN "parentHash" DROP NOT NULL;`);
+          queries.push(`UPDATE ${tableName} SET "parentHash" = NULL;`);
+          queries.push(
+            `CREATE UNIQUE INDEX IF NOT EXISTS "poi_parent_hash" ON ${tableName} ("parentHash") WHERE "parentHash" IS NOT NULL`
+          );
+        }
+      }
+
+      if (queries.length) {
+        for (const query of queries) {
+          try {
+            await this.poiRepo?.model.sequelize?.query(query, {type: QueryTypes.SELECT});
+          } catch (e) {
+            logger.error(`Migration poi failed with query: ${`query`}`);
+            throw e;
+          }
+        }
+        logger.info(`Successful migrate Poi`);
+      }
+      // Remove and Change column from sequelize not work, it only applies to public schema
+      // https://github.com/sequelize/sequelize/issues/13365
+      // await this.poiRepo?.model.sequelize?.getQueryInterface().changeColumn(tableName,'mmrRoot',{})
+    } catch (e) {
+      throw new Error(`Failed to migrate poi table}`);
+    }
+    // Before migration `latestSyncedPoiHeight` haven't been record in Db meta
+    // we try to find the first height from current poi table. and set for once
+    const genesisPoi = await this.poiRepo.getFirst();
+    if (genesisPoi && (genesisPoi.hash === null || genesisPoi.parentHash === null)) {
+      await this.createGenesisPoi(genesisPoi);
+    }
   }
 
-  private async fetchPoiBlockHashFromDb(): Promise<Uint8Array | null> {
-    const lastPoi = await this.poiRepo?.getLatestPoi();
-    if (lastPoi === null || lastPoi === undefined) {
-      return null;
-    } else if (lastPoi !== null && lastPoi.hash) {
-      return lastPoi.hash;
-    } else {
-      throw new Error(`Poi found but can not get latest hash`);
+  async ensureGenesisPoi(height: number) {
+    if (this._latestSyncedPoi) {
+      return;
     }
+    const poiBlock = await this.poiRepo.getPoiById(height);
+    if (poiBlock === undefined) {
+      throw new Error(`Ensure genesis poi failed, could not find poi ${height}`);
+    }
+    await this.createGenesisPoi(poiBlock);
   }
 
-  async getLatestPoiBlockHash(): Promise<Uint8Array> {
-    if (!this.latestPoiBlockHash || !isMainThread) {
-      const poiBlockHash = await this.fetchPoiBlockHashFromDb();
-      if (poiBlockHash === null || poiBlockHash === undefined) {
-        this.latestPoiBlockHash = DEFAULT_PARENT_HASH;
+  async syncPoi(exitHeight?: number, logging?: boolean): Promise<void> {
+    if (this.isSyncing) return;
+    this.isSyncing = true;
+    while (!this.isShutdown) {
+      if (!this._latestSyncedPoi) {
+        await delay(10);
```

Comment:
If _latestSyncedPoi is undefined, which means we haven't sync any Poi yet, no genesis poi been found.  It delay for genesis poi to be created, delay will stop after `ensureGenesisPoi`, https://github.com/subquery/subql/blob/bfe6de09f6743db219f5e46f2022bf41711feb3a/packages/node-core/src/indexer/poi/poi.service.ts#L160

Reply:
I don't think we should have random delays like this. Its near impossible to follow code with things like that. e.g. if for some reason `ensureGenesisPoi` fails or a refactor means its not called we could be scratching our heads for hours trying to figure out why this isn't running. 

We're better off calling functions in the correct order

Reply:
Genesis Poi need to create one time only. 
Once in metadata stored `latestSyncedPoi` and in the `init` will set this object, the sync process will immediately start. 

 Before I think we could make sure sync process start only after when the non-null operation block been processed, rather than checking every 10 seconds, but we also could face an issue here: 
Assume we restart the app, in current poi, latestSyncedPoi is 1000, and we have poi 1002,1005 awaiting to be synced. 
Then after 1005 ,there is no blocks with operation. which means 1002 and 1005 will be never been synced.. 

In current implementation will solve this issue. 

I will leave a comment with this method